pkg/connector/login.go

```diff
@@ -0,0 +1,311 @@
+package connector
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"go.mau.fi/whatsmeow"
+	"go.mau.fi/whatsmeow/store"
+	"maunium.net/go/mautrix/bridgev2"
+	"maunium.net/go/mautrix/bridgev2/database"
+
+	"maunium.net/go/mautrix-whatsapp/pkg/waid"
+)
+
+func (wa *WhatsAppConnector) GetLoginFlows() []bridgev2.LoginFlow {
+	return []bridgev2.LoginFlow{
+		{
+			Name:        "QR",
+			Description: "Scan a QR code to pair the bridge to your WhatsApp account",
+			ID:          "qr",
+		},
+		{
+			Name:        "Pairing code",
+			Description: "Input your phone number to get a pairing code, to pair the bridge to your WhatsApp account",
+			ID:          "pairing-code",
+		},
+	}
+}
+
+func (wa *WhatsAppConnector) CreateLogin(_ context.Context, user *bridgev2.User, flowID string) (bridgev2.LoginProcess, error) {
+
+	if flowID == "qr" {
+		return &QRLogin{User: user, Main: wa}, nil
+	} else if flowID == "pairing-code" {
+		return &PairingCodeLogin{User: user, Main: wa}, nil
+	} else {
+		return nil, fmt.Errorf("invalid login flow ID")
+	}
+}
+
+func makeQRChan(connector *WhatsAppConnector, user *bridgev2.User) (client *whatsmeow.Client, qrChan <-chan whatsmeow.QRChannelItem, cancelFunc context.CancelFunc, err error) {
+	log := connector.Bridge.Log.With().
+		Str("action", "login").
+		Stringer("user_id", user.MXID).
+		Logger()
+	qrCtx, cancel := context.WithCancel(log.WithContext(context.Background()))
+	cancelFunc = cancel
+
+	device := connector.DeviceStore.NewDevice()
+
+	wa := &WhatsAppClient{
+		Main:   connector,
+		Device: device,
+	}
+
+	wa.MakeNewClient(log)
+
+	client = wa.Client
+
+	qrChan, err = client.GetQRChannel(qrCtx)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	err = client.Connect()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	return
+}
+
+func makeUserLogin(ctx context.Context, user *bridgev2.User, client *whatsmeow.Client) (ul *bridgev2.UserLogin, err error) {
+	newLoginID := waid.MakeWAUserLoginID(client.Store.ID)
+	ul, err = user.NewLogin(ctx, &database.UserLogin{
+		ID:         newLoginID,
+		RemoteName: client.Store.PushName,
+		Metadata: &UserLoginMetadata{
+			WADeviceID: client.Store.ID.Device,
+		},
+	}, &bridgev2.NewLoginParams{
+		DeleteOnConflict: true,
+	})
+	return
+}
+
+type QRLogin struct {
+	User       *bridgev2.User
+	Main       *WhatsAppConnector
+	Client     *whatsmeow.Client
+	cancelChan context.CancelFunc
+	QRChan     <-chan whatsmeow.QRChannelItem
+
+	AccData *store.Device
+}
+
+var _ bridgev2.LoginProcessDisplayAndWait = (*QRLogin)(nil)
+
+func (qr *QRLogin) Cancel() {
+	qr.cancelChan()
+	go func() {
+		for range qr.QRChan {
+		}
+	}()
```

Comment:
It's a receiving channel, afaik you can only close sending or bidirectional ones -- and whatsmeow does not give us a function to cancel the qr channel -- (it can only be closed from the whatsmeow side)

Reply:
We are already canceling the context we used when we made the qr channel so it should be fine